block-producer/src/txqueue/mod.rs

```diff
@@ -0,0 +1,132 @@
+use std::{fmt::Debug, sync::Arc, time::Duration};
+
+use async_trait::async_trait;
+use tokio::{sync::RwLock, time};
+
+use crate::{batch_builder::BatchBuilder, SharedProvenTx, SharedRwVec};
+
+// TRANSACTION QUEUE
+// ================================================================================================
+
+#[derive(Debug)]
+pub enum VerifyTxError {}
+
+#[async_trait]
+pub trait TransactionVerifier: Send + Sync + 'static {
+    async fn verify_tx(
+        &self,
+        tx: SharedProvenTx,
+    ) -> Result<(), VerifyTxError>;
+}
+
+#[derive(Debug)]
+pub enum AddTransactionError {
+    VerificationFailed(VerifyTxError),
+}
+
+#[async_trait]
+pub trait TransactionQueue: Send + Sync + 'static {
+    async fn add_transaction(
+        &self,
+        tx: SharedProvenTx,
+    ) -> Result<(), AddTransactionError>;
+}
+
+// DEFAULT TRANSACTION QUEUE
+// ================================================================================================
+
+pub struct DefaultTransactionQueueOptions {
+    /// The frequency at which we try to build batches from transactions in the queue
+    pub build_batch_frequency: Duration,
+
+    /// The size of a batch
+    pub batch_size: usize,
+}
+
+pub struct DefaultTransactionQueue<BB: BatchBuilder, TV: TransactionVerifier> {
+    ready_queue: SharedRwVec<SharedProvenTx>,
+    tx_verifier: Arc<TV>,
+    batch_builder: Arc<BB>,
+    options: DefaultTransactionQueueOptions,
+}
+
+impl<BB, TV> DefaultTransactionQueue<BB, TV>
+where
+    TV: TransactionVerifier,
+    BB: BatchBuilder,
+{
+    pub fn new(
+        tx_verifier: Arc<TV>,
+        batch_builder: Arc<BB>,
+        options: DefaultTransactionQueueOptions,
+    ) -> Self {
+        Self {
+            ready_queue: Arc::new(RwLock::new(Vec::new())),
+            tx_verifier,
+            batch_builder,
+            options,
+        }
+    }
+
+    pub async fn run(self) {
+        let mut interval = time::interval(self.options.build_batch_frequency);
+
+        loop {
+            interval.tick().await;
+            self.try_build_batches().await;
+        }
```

Comment:
What you're suggesting is roughly [what I had in the previous PR](https://github.com/0xPolygonMiden/miden-node/blob/c6e5df9ebc9fbf97d07d1d71677ac1de952702f8/block-producer/src/tasks/txqueue/mod.rs#L153-L161). We decided against it because it added extra complexity for little benefit.

> The above is to show how 1 block of latency can be added.

Couple of comments:
+ We're thinking of this timer to be ~5ms, so the system is retrying ~50x/block time
  + Also, if proving a batch takes ~1s, the extra 5ms we wait before retrying is negligeable
+ This is currently the simplest possible conflict resolution strategy, and will become more sophisticated in the future (while still being timer-based).
+ The extra "wake up and poll" behavior when no new transactions came in the queue adds negligeable load

So we preferred this strategy as it achieves roughly the same outcome, but is much simpler. What do you think?

Reply:
> We decided against it because it added extra complexity for little benefit.

What is the rationale behind it? I haven't seen discussion on https://github.com/0xPolygonMiden/miden-node/pull/44

> We're thinking of this timer to be ~5ms

That is a fairly low waiting time. Are we sure we want to do that? Suppose we add logging into this loop, or that we want to attach a debugger to see what is happening, do we want to reach this every 5ms?